### DISCUSSION TITLE
Feasibility of ACEs

### DISCUSSION THREAD

**Discussion by lakaschus** [daveshap/ACE_Framework#8] [General]
Hey everyone!

I'm super excited about the ACE framework and can't wait to see more details in the upcoming papers. In the meantime, I would like to discuss the feasibility of implementing autonomous cognitive entities.

Has anyone tried to estimate how many model inferences we'd need for a project using ACE? And what about the costs? I think it'd be super helpful to get a sense of that.

I played a lot with similar tools like Auto-GPT, and it's very hard to get something out of them that vanilla ChatGPT cannot. One of the reasons are missing self-correction and reflection modules. From experience, it can take dozens of prompts to get the right answer for even simple subtasks. In the ACE framework you will have so much communication going on and each communication might have errors or "drifts" (i.e., changing goal, etc.), which would need to be reflected on and subsequently corrected. I feel like this will lead to an exponential number of model inferences using GPT-4 level LLMs, at least without human intervention.

What are your thoughts on the feasibility?

Really looking forward to seeing a concrete implementation!

**Comment by daveshap** [comment]
We're gonna figure that out with the MVP. Check out the roadmap here: https://github.com/daveshap/ACE_Framework/blob/main/agile.md

**Comment by mikeslattery** [comment]
> I played a lot with similar tools like Auto-GPT, and it's very hard to get something out of them ... One of the reasons are missing self-correction and reflection modules. From experience, it can take dozens of prompts to get the right answer for even simple subtasks.

Yeah.   This is *the* issue ACE will have to solve; staying on task and successfully breaking problems down well enough to be solvable.   So far nothing has come close, really.   LLMs and agents are helpful in the small, but get distracted when you scale and/or make them autonomous.

My use case has been code gen.  Before ACE, I had given up on the idea of autonomous codegen.  I concluded we need a GPT-5, or even better (Gemini?). [GPT Pilot](https://github.com/Pythagora-io/gpt-pilot/)  is the only agent I've found that's come close to being practical, mainly because it depends on a lot of manual intervention.   The only reason I have any optimism is because of Shapiro's involvement and the thoughtfulness of the design.

There's been a ton of research on techniques that work well, yet no single agent (that I've seen) employs them well or at all (e.g. reflexion, graph of thought, APE, etc).   Maybe ACE will be the first robust agent that makes this seemingly impossible goal possible.

**Reply by daveshap** [reply]
The problem you're talking about is called "cognitive control" which was one of the chief problems I worked to solve.

**Comment by DataBassGit** [comment]
Something to keep in mind, cost of inference is dropping drastically. While compute is still king, I think we're going to start heading towards models that are much smaller and rely on larger context windows and better attention mechanisms. There's no need to train a model on the entire internet. "Textbooks are all you need."

Look back into studies on in-context learning. If you provide an LLM with a question+all of the information that is needed to answer that question, it can answer it even when it was never trained on the data. This data can actually come straight from the internet, or be in a database. Additional relevant data can be discerned using knowledge graphs, so when you ask about stars, the model gets context related to stars, but also context related to physics and gravity.

The reason why I think we will go this way (Unless we pivot to a new model that can easily and cheaply be updated) is because information has a shelf life. Let's take python for example. The basic syntax remains relatively stable, but new libraries come out all the time. Old libraries get updated. This is one of the major sources of glitchy code from ChatGPT. However, it's completely able to write code using libraries that were released after it was trained if you provide it the library documentation first.

Ultimately, the amount of inference costs are going to depend on what your application is. A system that incorporates vision is going to be more expensive, perhaps prohibitively expensive, since every image will need to be processed. If you build it in a way that loops, like an autonomous agent that you give an objective to, you have to worry about the size of your prompts as well as the number of prompts. 100 small prompts could easily be less expensive than 10 large prompts.

But we already have autonomous agents. What this stack adds is control, alignment, and interpretability. A lot of that comes without inference costs, but some of it does add additional tokens to your bill. But not all of it has to go through GPT. The idea is to allow for any message to go to any AI system, since they all take text as an input. That means you might need multiple API keys, or you may need to host open source models in the cloud. You can even train a model specifically for alignment verification (Morality/Ethics layer) that is completely separate from the agent layer so that it can't be tampered with.

But the main thing I wanted to draw attention to was, Pretty soon, cost won't matter.

**Reply by j-loquat** [reply]
It would be interesting to see how quickly a project like ChatDev https://github.com/OpenBMB/ChatDev could adopt this ACE framework. Many pieces already in place. Thoughts?
